pkg/sanity/controller.go

```diff
@@ -69,1100 +69,1014 @@ func isControllerCapabilitySupported(
-
-		req := &csi.CreateVolumeRequest{}
-
-		if sc.Secrets != nil {
-			req.ControllerCreateSecrets = sc.Secrets.CreateVolumeSecret
-		}
+	Describe("GetCapacity", func() {
+		BeforeEach(func() {
+			if !isControllerCapabilitySupported(c, csi.ControllerServiceCapability_RPC_GET_CAPACITY) {
+				Skip("GetCapacity not supported")
+			}
+		})
 
-		_, err := c.CreateVolume(context.Background(), req)
-		Expect(err).To(HaveOccurred())
+		It("should return capacity (no optional values added)", func() {
+			_, err := c.GetCapacity(
+				context.Background(),
+				&csi.GetCapacityRequest{})
+			Expect(err).NotTo(HaveOccurred())
 
-		serverError, ok := status.FromError(err)
-		Expect(ok).To(BeTrue())
-		Expect(serverError.Code()).To(Equal(codes.InvalidArgument))
+			// Since capacity is int64 we will not be checking it
+			// The value of zero is a possible value.
+		})
 	})
 
-	It("should fail when no volume capabilities are provided", func() {
+	Describe("ListVolumes", func() {
+		BeforeEach(func() {
+			if !isControllerCapabilitySupported(c, csi.ControllerServiceCapability_RPC_LIST_VOLUMES) {
+				Skip("ListVolumes not supported")
+			}
+		})
 
-		req := &csi.CreateVolumeRequest{
-			Name: "name",
-		}
+		It("should return appropriate values (no optional values added)", func() {
+			vols, err := c.ListVolumes(
+				context.Background(),
+				&csi.ListVolumesRequest{})
+			Expect(err).NotTo(HaveOccurred())
+			Expect(vols).NotTo(BeNil())
 
-		if sc.Secrets != nil {
-			req.ControllerCreateSecrets = sc.Secrets.CreateVolumeSecret
-		}
+			for _, vol := range vols.GetEntries() {
+				verifyVolumeInfo(vol.GetVolume())
+			}
+		})
 
-		_, err := c.CreateVolume(context.Background(), req)
-		Expect(err).To(HaveOccurred())
+		// TODO: Add test to test for tokens
 
-		serverError, ok := status.FromError(err)
-		Expect(ok).To(BeTrue())
-		Expect(serverError.Code()).To(Equal(codes.InvalidArgument))
+		// TODO: Add test which checks list of volume is there when created,
+		//       and not there when deleted.
 	})
 
-	It("should return appropriate values SingleNodeWriter NoCapacity Type:Mount", func() {
-
-		By("creating a volume")
-		name := "sanity"
+	Describe("CreateVolume", func() {
+		BeforeEach(func() {
+			if !isControllerCapabilitySupported(c, csi.ControllerServiceCapability_RPC_CREATE_DELETE_VOLUME) {
+				Skip("CreateVolume not supported")
+			}
+		})
 
-		req := &csi.CreateVolumeRequest{
-			Name: name,
-			VolumeCapabilities: []*csi.VolumeCapability{
-				{
-					AccessType: &csi.VolumeCapability_Mount{
-						Mount: &csi.VolumeCapability_MountVolume{},
-					},
-					AccessMode: &csi.VolumeCapability_AccessMode{
-						Mode: csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER,
+		It("should fail when no name is provided", func() {
+			_, err := c.CreateVolume(
+				context.Background(),
+				&csi.CreateVolumeRequest{
+					ControllerCreateSecrets: sc.Secrets.CreateVolumeSecret,
+				},
+			)
+			Expect(err).To(HaveOccurred())
```

Comment:
If CreateVolume actually succeeded, then you still need to cleanup the volume.  Same for all negative tests.

Reply:
Yes, I had asked myself about that too, but then dismissed it as unlikely. Will add it.

Reply:
Done, see d34fca7.